r/src/array.cpp

```diff
@@ -828,36 +849,21 @@ SEXP TimeArray_to_Vector(const std::shared_ptr<Array>& array, int32_t multiplier
 
 SEXP DecimalArray(const std::shared_ptr<Array>& array) {
   auto n = array->length();
-  NumericVector vec(no_init(n));
-
-  if (n == 0) return vec;
-
-  auto null_count = array->null_count();
-  if (null_count == n) {
-    std::fill(vec.begin(), vec.end(), NA_REAL);
-    return vec;
-  }
+  ComplexVector vec(no_init(n));
 
-  auto p_vec = reinterpret_cast<double*>(vec.begin());
-  const auto& decimals_arr =
-      internal::checked_cast<const arrow::Decimal128Array&>(*array);
-
-  if (array->null_count()) {
-    internal::BitmapReader bitmap_reader(array->null_bitmap()->data(), array->offset(),
-                                         n);
+  auto p_vec = reinterpret_cast<arrow::Decimal128*>(vec.begin());
+  auto p_values = GetValuesSafely<arrow::Decimal128>(array->data(), 1, array->offset());
 
-    for (size_t i = 0; i < n; i++, bitmap_reader.Next()) {
-      p_vec[i] = bitmap_reader.IsNotSet()
-                     ? NA_REAL
-                     : std::stod(decimals_arr.FormatValue(i).c_str());
-    }
-  } else {
-    for (size_t i = 0; i < n; i++) {
-      p_vec[i] = std::stod(decimals_arr.FormatValue(i).c_str());
-    }
-  }
+  // TODO: deal with NA
+  std::copy_n(p_values, n, p_vec);
 
-  return vec;
+  List res = List::create(_["data"] = vec);
+  res.attr("class") =
+      CharacterVector::create("arrow_decimal128", "vctrs_rcrd", "vctrs_vctr");
+  res.attr("scale") = static_cast<arrow::Decimal128Type*>(array->type().get())->scale();
+  res.attr("precision") =
+      static_cast<arrow::Decimal128Type*>(array->type().get())->precision();
```

Comment:
Should use `checked_cast` and a const reference here to get a better error message if this fails, in debug mode. `checked_cast` compiles into a `static_cast` in release mode builds.

Reply:
Something like

```c++
checked_cast<const arrow::Decimal128Type&>(*array->type())->precision()
```